rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -57,44 +59,87 @@ def replication_enabled(self):
         """Check whether replication enabled."""
         return ENVIRONMENT.get_value("REPLICATION_TO_RELATION_ENABLED", default=False, cast=bool)
 
-    def generate_relations_from_current_state_of_role(self):
+    def get_current_role_relations(self):
+        """Get current roles relations."""
+        return self.current_role_relations
+
+    def load_relations_from_current_state_of_role(self):
         """Generate relations from current state of role and UUIDs for v2 role and role binding from database."""
         if not self.replication_enabled():
             return
         try:
             logger.info(
                 "[Dual Write] Generate relations from current state of role(%s): '%s'", self.role.uuid, self.role.name
             )
-            relations = migrate_role(
-                self.role, False, str(self.root_workspace.uuid), self.org_id, True, True, True, False
+
+            try:
+                self.binding_mapping = self.role.binding_mapping
+            except BindingMapping.DoesNotExist:
+                logger.warning(
+                    "[Dual Write] Binding mapping not found for role(%s): '%s'. Creating new binding mapping.",
+                    self.role.uuid,
+                    self.role.name,
+                )
```

Comment:
This was not getting hit before for how the test cases were set up, but now this case is exposed so I added this warning instead of failing. Whether or not we fail this depends how we want to handle migration. I don't actually think allowing this will cause any inconsistency to happen as long as if there are no mappings stored in rbac, it also means there are no corresponding relations in the Relations API. So I think a warning is appropriate.

Reply:
I think we could get rid of the nested `try/except` here and just catch the error to log the warning below, which might make it easier to read?

```python
try:
    logger.info(
        "[Dual Write] Generate relations from current state of role(%s): '%s'", self.role.uuid, self.role.name
    )

    self.binding_mapping = self.role.binding_mapping

    relations, _ = migrate_role(
        self.role,
        write_relationships=False,
        root_workspace=str(self.root_workspace.uuid),
        default_workspace=self.org_id,
        current_bindings=self.binding_mapping,
    )

    self.current_role_relations = relations
except Exception as e:
    raise DualWriteException(e)
except BindingMapping.DoesNotExist:
    logger.warning(
        "[Dual Write] Binding mapping not found for role(%s): '%s'. Creating new binding mapping.",
        self.role.uuid,
        self.role.name,
    )
```

Reply:
I do like how that reads better. One problem is that we still need to run `migrate_role` in that case though. Wondering if there is another way 🤔

Reply:
Maybe we don't actually? 🤔

Reply:
Yeah I think this should work since in this case there _shouldn't_ be any relations that we have to remove. I just had to put the generic `Exception` catch block last. Good one!